### DISCUSSION TITLE
[feature] add a tplHelm function to use the helm engine templating feature

### DISCUSSION THREAD

**Discussion by Diliz** [helmfile/helmfile#911] [Ideas]
Hello!

There's a lot of differences between the helmfile templating engine and the helm one, one of my main issue with helmfile was to not be able to use helm templating features in the helmfile.yaml file.

I'm actually templating values.yaml files coming from helm charts, the issue I have here is that I have an issue with values.yaml files that contains helm templating in it, like the include things, I actually made it work.

Just so you have some context, I'm fetching the values files needed, merging it, then inject the values I fetched and merged to the fetched and merged values template itself with the tpl function, as follow:
```
###
### Import Values Files and Merge it
###

{{- $files := dict "paths" (list "values.yaml") -}}
{{- range $_,$valuesFilePath := readDir "." -}}
  {{- $chunks := splitList "-" $target -}}
  {{- range $index,$_ := $chunks -}}
    {{- $validValueFile := printf "values-%s" (join "-" (slice $chunks 0 (add $index 1))) -}}
    {{- if contains $validValueFile $valuesFilePath -}}
      {{- $_ := set $files "paths" (append $files.paths $valuesFilePath) -}}
    {{- end -}}
  {{- end -}}
{{- end -}}

{{- $mergedValues := dict "Values" dict -}}
{{- range $_,$valuesFilePath := $files.paths -}}
  {{- $valuesFile := readFile $valuesFilePath | fromYaml -}}
  {{-
    $_ := set
      $mergedValues "Values" (
        (deepCopy $mergedValues.Values)
        | mustMerge (deepCopy $valuesFile)
      )
  -}}
{{- end -}}

{{-
  $_ := set $mergedValues "Values" (
      dict "global" $global | mustMerge (deepCopy $mergedValues.Values)
    )
-}}

{{- $mergedValuesYaml := $mergedValues.Values | toYaml -}}
{{- $valuesYaml := $mergedValues | tpl $mergedValuesYaml }}
{{- $values := $valuesYaml | fromYaml }}
```

Doing the same with the overlays if there's one:
```
###
### Import Kustomizations and merge it
###

{{- $defaultOverlay := readFile "kustomization.yaml" | fromYaml -}}

{{- $strategicMergePatches := ($defaultOverlay | get "strategicMergePatches" list) | uniq -}}
{{- $patchs := concat ($defaultOverlay | get "patchs" list) ($defaultOverlay | get "jsonPatches" list) | uniq -}}
{{- $resources := $defaultOverlay | get "resources" list -}}

{{-
  $mergedOverlay :=
    dict
      "strategicMergePatches" $strategicMergePatches
      "jsonPatches" $patchs
      "resources" $resources
-}}

{{- $mergedOverlayYaml := (deepCopy $mergedOverlay) | toYaml -}}
{{- $overlayYaml := $mergedValues | tpl $mergedOverlayYaml -}}
{{- $overlay := $overlayYaml | fromYaml -}}
```

The thing is, if you try to use an helm chart like this one: https://github.com/jfrog/charts/blob/master/stable/xray/values.yaml#L151

It will not work because the helm engine functions are not available here.

So, I currently have two possibilities:
- use the helm engine in some way at some point to have the possibility to use helm functions
- the other is to have a try/catch function, and just not template the values file if it's not working, still, I don't know how to do that, if doable at all.

Do you have any recommendations with this? Is it something implementable?

Thanks for reading,
Diliz

**Comment by yxxhero** [comment]
@Diliz PR is welcome.

**Reply by Diliz** [reply]
Now that I read myself again, I think the include function is a no go anyway, depending in which template it is called, if the include is used in a subchart, or a subsubchart, the values it should take would be something like this:
```
subchartAlias:
   valuesNeededForTheInclude: ....
  subsubchartAlias:
     valuesNeededForTheInclude: ....
```

A simple tpl function would not work, and this is probably why helm itself is not templating the values.yaml files to begin with, I assume that some functions are not simply usable in this kind of thing.

So a try/catch feature would be better to handle at least something like this, but it would need to catch the error, etc...

**Comment by JuryA** [comment]
Yes, personally I believe that expanding the template engine itself to be more compatible with how Helm handles it is a crucial necessity - let me explain why. Boldly, I dare to elaborate on the idea and, in my opinion, take it to a completely different level: I can definitely envision native support for Helm library charts and calling their named templates in Helmfile... That would be a very beneficial gamechanger, don't you think? 😎

**Comment by Diliz** [comment]
Humm, can be cool, but not really simple to do, having all the functions that helm implements natively would mean to redo the dependency and subdependency tree beforehand, if you take the jfrog xray or platform helm chart as an exemple, they use the tpl function in their templates, they use include function as well and define some named templates, which is, in an umbrella helm chart, really hard to manage, if multiple same named templates comes from different charts, it becomes really troublesome.

One "simple" way to "fix" this would be to split the parent values, isolate the values for the subchart, fetch the subchart, load the named templates only for the subchart, inject the specific subchart values in it, which you can template if you do that, etc, if a subsubchart is present, repeat the same process with the subchart as the parent.

But even with that, it would mean harder templated helmfile, if you calculate the chart name or namespace depending on some environment values or else, it'll be troublesome to handle

Edit: still, having helm applications/libraries named templates available for include in the helmfile could be cool, and being able to overwrite a named template would be reaaaaaally good

**Reply by JuryA** [reply]
I don't think the implementation would be that complex. In both cases, **Go Templates** are used; the difference lies only in the basic "presetting." In my opinion, the easiest approach would be to "adapt" the Go package **https://pkg.go.dev/helm.sh/helm/v3/pkg/engine** and aim for maximum standardization of **Go Template** functions, similar to

https://github.com/helm/helm/blob/f32a527a060157990e2aa86bf45010dfb3cc8b8d/pkg/engine/funcs.go#L43-L76

(`include`, `tpl`, `required`, `fail`, ... are added here https://github.com/helm/helm/blob/f32a527a060157990e2aa86bf45010dfb3cc8b8d/pkg/engine/engine.go#L116-L210)

I believe this is a feasible task. From there, it's just a small step towards implementing **Helm Library Charts**, using the `Render` function - see: 

https://github.com/helm/helm/blob/f32a527a060157990e2aa86bf45010dfb3cc8b8d/pkg/engine/engine.go#L75-L78

@mumoshu @yxxhero What do you think?

**Reply by JuryA** [reply]
For illustration purposes, I am attaching one of our advanced implementations of **Helmfile Go Templates**:

```mustache
{{- /* Try to get `PWD` from: 1) `.Values._helmfile.PWD`, 2) EnvVar `PWD`, 3) root of Git repo (of this Helmfile) (fallback) */ -}}
{{ $PWD := (coalesce (.Values | getOrNil "_helmfile.PWD") (env "PWD") (exec "git" (list "rev-parse" "--show-toplevel") | trim)) }}
{{ $helmfileDir := .Values | get "_helmfile.helmfileDir" "../" }}
{{ $envs := .Values | get "_helmfile.envs" "../" }}
{{ $envFiles := (dict) }}
environments:
  "{{ .Environment.Name }}":
    values:
      - _helmfile:
          envFiles:
          {{- /* Check if the environment name contains ".env.yaml" */}}
          {{ if contains ".env.yaml" .Environment.Name }}
            {{- /* Clean the environment name */}}
            {{ $path := clean .Environment.Name }}
            {{- /* If the path is not absolute, resolve it using $PWD */}}
            {{ if not (osIsAbs $path) }}
              {{ $path = exec "readlink" (list "-f" (join "/" (list $PWD $path))) | trim }}
            {{ end }}
            {{- /* Indent and output the environment name along with its contents converted to JSON */}}
            {{- .Environment.Name | quote | indent 14 }}: {{ (readFile $path | fromYaml) | toJson }}
          {{ else }}
            {{- /* Loop through the files in the "envs" directory and look for ".env.yaml" files */}}
            {{ if $envs }}
            {{ range $item := readDir $envs }}
              {{- if contains ".env.yaml" $item }}
                {{- /* Read the file contents and convert them to YAML */}}
                {{- $_item := (readFile $item | fromYaml) -}}
                {{- /* Add a description to the environment file indicating that it's built-in */}}
                {{- $_item := set $_item "~description" (join " " (list "(environments)" ($_item | getOrNil "~description"))) -}}
                {{- /* Extract the base name of the file and indent it, then output its contents converted to JSON */}}
                {{- $item | splitList ".env.yaml" | first | osBase | quote | indent 14 }}: {{ $_item | toJson }}
              {{ end }}
            {{ end }}
            {{ end }}
            {{- /* Loop through the files in the current directory and look for ".env.yaml" files */}}
            {{ range $item := readDir $PWD }}
              {{- if contains ".env.yaml" $item }}
                {{- /* Extract the base name of the file and indent it, then read its contents and convert them to YAML */}}
                {{- $item | splitList ".env.yaml" | first | osBase | quote | indent 14 }}: {{ (readFile $item | fromYaml) | toJson }}
              {{ end }}
            {{ end }}
          {{ end }}
```

**Reply by JuryA** [reply]
And this one:

```mustache
# Try to get `PWD` from: 1) `.Values._helmfile.PWD`, 2) EnvVar `PWD`, 3) root of Git repo (of this Helmfile) (fallback)
{{ $PWD := (coalesce (.Values | getOrNil "_helmfile.PWD") (env "PWD") (exec "git" (list "rev-parse" "--show-toplevel") | trim)) }}
{{ $envFile := dig .Environment.Name (dict) (.Values | get "_helmfile.envFiles") }}
environments:
  default: {}

  # Check if `.Environment.Name` is set and is not "default" (in other words if helmfile is executed with flag
  # `-e`/`---environment`).
  {{ if ne .Environment.Name "default" }}
    {{- if empty $envFile -}}
      {{- fail (printf "\n\nERROR: [%s]: Environment not found! Cannot load [%s.env.yaml] file." .Environment.Name .Environment.Name ) -}}
    {{- end -}}
  # Create an helmfile `environment` with the name from `.Environment.Name`, which must be defined. Therefore helmfile
  # must be executed with flag `-e`/`---environment` set to valid value (this will set `.Environment.Name`). This value
  # is valid only if appropriate file with name `<value>.env.yaml` exists in the root of actual Git repository.
  "{{ .Environment.Name }}":
    values:
      # Search for appropriate `.env.yaml` in the root of actual Git repository and load it to environment
      # `StateValues`. If such file doesn't exist, helmfile process fails and exits.
      - {{ $envFile | toJson }}
{{- end -}}
```

**Reply by JuryA** [reply]
@mumoshu This is an example showcasing how we handle **environments** in a rather unconventional way, as discussed in https://github.com/helmfile/helmfile/issues/913#issuecomment-1630894826 ... 😎
